### PR TITLE
ddl, planner, config: report unsupported for create/drop temporary table

### PR DESCRIPTION
**Pull request by ffutop**
### What problem does this PR solve? 

Fix https://github.com/pingcap/parser/issues/609
report unsupported error.

### What is changed and how it works?

```bash
# Turn off option ignore-ddl-temporary-keyword
mysql> CREATE TEMPORARY TABLE t (a INT);
ERROR 8200 (HY000): 'CREATE TEMPORARY TABLE' is currently unsupported
mysql> DROP TEMPORARY TABLE t;
ERROR 8200 (HY000): 'DROP TEMPORARY TABLE' is currently unsupported
mysql> CREATE TABLE t (a INT);
Query OK, 0 rows affected (0.01 sec)

mysql> DROP TABLE t;
Query OK, 0 rows affected (0.01 sec)

mysql> CREATE TEMPORARY SEQUENCE seq START WITH 100;
ERROR 8200 (HY000): 'CREATE TEMPORARY SEQUENCE' is currently unsupported
mysql> DROP TEMPORARY SEQUENCE seq;
ERROR 8200 (HY000): 'DROP TEMPORARY SEQUENCE' is currently unsupported
mysql> CREATE SEQUENCE seq START WITH 100;
Query OK, 0 rows affected (0.00 sec)

mysql> DROP SEQUENCE seq;
Query OK, 0 rows affected (0.00 sec)

```

```bash
# Turn on option ignore-ddl-temporary-keyword (default)
mysql> CREATE TEMPORARY TABLE t (a INT);
Query OK, 0 rows affected, 1 warning (0.01 sec)

mysql> SHOW CREATE TABLE t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.00 sec)

mysql> DROP TEMPORARY TABLE t;
Query OK, 0 rows affected, 1 warning (0.01 sec)

mysql> SHOW TABLES;
Empty set (0.00 sec)

mysql> CREATE TEMPORARY SEQUENCE seq START WITH 100;
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> SHOW CREATE TABLE seq\G
*************************** 1. row ***************************
       Table: seq
Create Table: CREATE TABLE `seq` (

) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.00 sec)

mysql> DROP TEMPORARY SEQUENCE seq;
Query OK, 0 rows affected, 1 warning (0.00 sec)
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by wwar** [comment]
May I suggest an error message similar to "CREATE TEMPORARY TABLE is currently unsupported".

The message "Unsupported temporary table" is ambiguous; it could be interpreted as temporary tables are supported, but this particular one is not.

**Comment by bb7133** [comment]
Hi @ffutop , thanks for your contribution!

**Comment by bb7133** [comment]
The implementation looks fine for me. However, changing such behaviors means a potential breakdown for TiDB users. For example, many users switched their databases from MySQL to TiDB may have their scripts with `TEMPORARY TABLE`, that's why TiDB ignores this keyword by design.

Would you please add a configuration option(naming 'compatible mode', for example)? You can check the implementation of https://github.com/pingcap/tidb/blob/master/config/config.toml.example#L61 as a reference.

**Comment by wwar** [comment]
@bb7133 I think the behavior proposed here is an improvement since the case where the keyword `TEMPORARY` can be ignored is quite narrow:
* The script would have to be single-threaded, and cleanup after itself to prevent name collisions.
* There are also potential leakage scenarios, since the script may not have assumed that another user could read the data, or assume concurrent usage.
* `TEMPORARY` tables also have a weird behavior where they can have the same name as a regular table, and reading from the temp table will take precedence.

**Comment by ffutop** [comment]
@bb7133 I'd like to take your advice. I'm not sure is it ok naming new option `ignore-ddl-temporary-keywords`

File: planner/core/preprocess_test.go [hunk 1]
```diff
@@ -206,6 +206,9 @@ func (s *testValidatorSuite) TestValidator(c *C) {
 		{"CREATE TABLE t1 (id INT NOT NULL, c1 VARCHAR(20) AS ('foo') VIRTUAL KEY NULL, PRIMARY KEY (id));", false, core.ErrUnsupportedOnGeneratedColumn},
 		{"CREATE TABLE t1 (id INT NOT NULL, c1 VARCHAR(20) AS ('foo') VIRTUAL KEY NOT NULL, PRIMARY KEY (id));", false, core.ErrUnsupportedOnGeneratedColumn},
 		{"create table t (a DOUBLE NULL, b_sto DOUBLE GENERATED ALWAYS AS (a + 2) STORED UNIQUE KEY NOT NULL PRIMARY KEY);", false, nil},
+
+		{"CREATE TEMPORARY TABLE t (a INT);", false, ddl.ErrUnsupportedTempTable.GenWithStack("'CREATE TEMPORARY TABLE' is currently unsupported")},
```
Comment (tangenta) [inline planner/core/preprocess_test.go]:
```suggestion
		{"CREATE TEMPORARY TABLE t (a INT);", false, ddl.ErrUnsupportedTempTable},
```

File: planner/core/preprocess_test.go [hunk 2]
```diff
@@ -206,6 +206,9 @@ func (s *testValidatorSuite) TestValidator(c *C) {
...
+		{"DROP TEMPORARY TABLE t;", false, ddl.ErrUnsupportedTempTable.GenWithStack("'DROP TEMPORARY TABLE' is currently unsupported")},
```
Comment (tangenta) [inline planner/core/preprocess_test.go]:
```suggestion
		{"DROP TEMPORARY TABLE t;", false, ddl.ErrUnsupportedTempTable},
```

**Comment by bb7133** [comment]
> @bb7133 I'd like to take your advice. I'm not sure is it ok naming new option ignore-ddl-temporary-keywords

How do you think about this option? @tangenta @kennytm

**Comment by kennytm** [comment]
1. Please also handle `CREATE TEMPORARY SEQUENCE` and `DROP TEMPORARY SEQUENCE`
2. Do we care about the `CREATE TEMPORARY TABLES` privilege
3. Having an option is fine. But are we going to keep this option for a long time (e.g. what would we do after we've actually implemented TEMPORARY TABLE)?

**Comment by ffutop** [comment]
@kennytm I can not find any mysql docs that introduce `CREATE/DROP SEQUENCE`.
And I tried following commands based on [parser.y]( https://github.com/pingcap/parser/blob/master/parser.y)

`CREATE SEQUENCE` create a new table. And `DROP SEQUENCE` not works.
 
```
tidb> CREATE SEQUENCE seq;
Query OK, 0 rows affected (0.00 sec)

tidb> SHOW TABLES;
+----------------+
| Tables_in_test |
+----------------+
| seq            |
+----------------+
1 row in set (0.00 sec)

tidb> DROP SEQUENCE seq;
Query OK, 0 rows affected (0.00 sec)

tidb> SHOW TABLES;
+----------------+
| Tables_in_test |
+----------------+
| seq            |
+----------------+
1 row in set (0.00 sec)

tidb> DROP TABLE seq;
Query OK, 0 rows affected (0.01 sec)

tidb> SHOW TABLES;
Empty set (0.00 sec)

```

**Comment by kennytm** [comment]
@ffutop You could refer to MariaDB's documentation https://mariadb.com/kb/en/create-sequence/ for the syntax (this mostly equivalent to the "sequence generator" feature in ANSI SQL:2003, also shared by PostgreSQL, Oracle Database, Microsoft Transact-SQL, IBM DB2, and probably many other RDBMS).

Anyway, my comment is that `CREATE TEMPORARY SEQUENCE seq ...` should have the same treatment as `CREATE TEMPORARY TABLE tbl (...)` regarding the "TEMPORARY" keyword. Since this feature was recently implemented I wasn't sure if it already works on latest TiDB, and it seems to be working.

**Comment by bb7133** [comment]
> I can not find any mysql docs that introduce CREATE/DROP SEQUENCE.

The support of `SEQUENCE` is in-progress(although it is not supported by MySQL, as what you find out already).

**Comment by ffutop** [comment]
@bb7133 PTAL

File: config/config.toml.example [hunk 3]
```diff
@@ -105,6 +105,11 @@ max-server-connections = 4096
 # Whether new collations are enabled, as indicated by its name, this configuration entry take effect ONLY when a TiDB cluster bootstraps for the first time.
 new_collations_enabled_on_first_bootstrap = false
 
+# ignore-ddl-temporary-keyword is used to make ddl TEMPORARY keyword compatible with previous version TiDB.
+# turn on this option, 'TEMPORARY' will ignored by TiDB as previous. Please make sure you fully understand the negative impact.
+# turn off this option, use 'CREATE/DROP TEMPORARY TABLE' or 'CREATE/DROP TEMPORARY SEQUENCE' will reported an error.
+ignore-ddl-temporary-keyword = false
```
Comment (zz-jason) [inline config/config.toml.example]: Consider an older version of TiDB, which do not have this config item in the config file, updated to this version. The default value of this item is `false`, then there is chance that the application may failed to run on the newer version TiDB.

BTW, it makes user confused when we using double negation. How about renaming it to `reserve-ddl-temporary-keyword`?

Comment (ffutop) [reply config/config.toml.example]: OK, I will change the `defaultConf.IgnoreDDLTemporaryKeyword` to `true`.

but, I cannot agree to your  renaming advice. `reserve-ddl-temporary-keyword` is  ambiguous compare with `ignore-xxx` for new user, `TEMPORARY` is standard SQL syntax, why need to reserve? set to `false` might be interpreted as

1.  no  `TEMPORARY` syntax supported for tidb, or 
2. just ignore `TEMPORARY`

**Comment by bb7133** [comment]
Hi @ffutop , please update the PR, thank you!

**Pull request closed without merging**